### ISSUE TITLE
Better Documentation regarding Size and Tree-Shaking

### ISSUE THREAD

**Issue by noahgrant** [moment/luxon#854]
Hello!

This follows from from https://github.com/moment/luxon/issues/703 and https://github.com/moment/luxon/issues/94#issuecomment-346481639.

So, I have just moved from moment to luxon for the reasons listed [on the moment homepage](https://momentjs.com/docs). When I read about using native APIs in more modern libraries or new versions of old libraries, I guess I assume that the new library will have a size savings, as well. Imagine my surprise, then, when I finally migrate everything over and find out that my `vendor.js` production-built file size has _increased_ by 170kB minified and 53kB gzipped!!

Had I known this ahead of time, I probably would've stuck with moment or moved to another library like Day.js (I still might). I'm currently not using moment's locales or its timezone addition, and at another company, I provided my own short-list of timezone data to keep moment's size down. In a similar manner, I'm not using the bulk of what Luxon exports (ie `Interval` and `Zone`), but since it's not modular I can't tell webpack, for example, to leave certain modules out. 

I've read discussions on things people have tried in attempt to make their file sizes smaller, but what I'd really like is a consensus on what to do for each of the popular build tools/environments, and have that be listed in the Luxon documentation page. For example, how can I point my build to use the smaller `global-es6/luxon.min.js` instead of `cjs-browser/luxon.min.js` that it's currently using?

Additionally, I think it'd be really helpful to be upfront about the size costs of using Luxon (especially that you are comparing it with the fully-loaded moment library). One ideal place would be in the [moment homepage](https://momentjs.com/docs) under the `Reasons to keep using Moment` section that discusses Luxon—knowing that library size does not actually get smaller but rather increases would be really helpful to know when comparing alternatives.

Thank you for reading!!

**Comment by noahgrant** [comment]
I also have a follow-up question I'd really like to understand: _why_ is Luxon so big?

If it utilizes native Intl APIs for locales and needs no timezone data, why is it so big?

**Comment by tcsaddul** [comment]
170kb is big in 2021? How about utilizing the browser cache with versioning? I guess it is negligible.

**Comment by lephro** [comment]
170Kb would be a lot... but 24 Kb is ok.

@tcsaddul While 170Kb doesnt sound like a lot, it actually matter once things start to add up. If every libary u use would increase its size by 170Kb, you would increase the loadings times by a lot... even in 2021. In worst case, you can loose a lot of users, just because of increased loading times. It depends on the project ofc, but overall its not something that you shoulnd't care about, if you're creating a website on a professional level.

**Comment by MarcelCoding** [comment]
If you import using `import { ... } from 'luxon/src/luxon';` instead of `import { ... } from 'luxon';`, the size will be about 50Kb smaller. This also works with `@types/luxon`.

**Comment by MarcelCoding** [comment]
@9mm What do you mean with an example? A real code example with size difference between the tow types of imports? Or what do you mean?

**Comment by 9mm** [comment]
Maybe @icambron could help to clarify some things?

**Comment by 9mm** [comment]
I will definitely say the way luxon handles timezones are A++. After using something like date-fns-tz which makes me die inside.

**Comment by icambron** [comment]
I've long had on my todo list to simply experiment with this, and perhaps make changes that enhance the tree shakability of the lib. Unfortunately, I just haven't gotten to it

**Comment by Bessonov** [comment]
I'm not sure about the reason, but it seems like luxon isn't tree-shakeable. It can be wrong build options or the way how library is written - in fluent-api way.

Tried with:

```typescript
import { DateTime } from 'luxon' // the only import from luxon
import { merge } from 'remeda' // the only import from remeda

[...]
		this.expire = DateTime.now().plus({
			seconds: expires_in - 5 // slack
		})
[...]
		console.log(merge({}, {}))
		if (DateTime.now() > this.expire) {
			return true
		}
```

It seems like it works for `merge`:
```javascript
function rr(t,e,n){var r=t.length-e.length,s=Array.from(e);if(0===r)return t.apply(void 0,s);if(1===r){var i=function(e){return t.apply(void 0,function(){for(var t=0,e=0,n=arguments.length;e<n;e++)t+=arguments[e].length;var r=Array(t),s=0;for(e=0;e<n;e++)for(var i=arguments[e],a=0,o=i.length;a<o;a++,s++)r[s]=i[a];return r}([e],s))};return(n||t.lazy)&&(i.lazy=n||t.lazy,i.lazyArgs=e),i}throw new Error("Wrong number of arguments")}

[...]

(console.log(function(){return rr(sr,arguments)}({},{})),er.now()>this.expire)
```
and it's not shown in the source-map-explorer. At least I can't find it without looking into the source map deeper. But at the core, there is no other code from remeda.

On the other side, luxon, give me this picture:
![image](https://user-images.githubusercontent.com/182451/119348588-5eb8f880-bc9d-11eb-8cb7-377c4e266308.png)


If I use
```typescript
import { Duration } from 'luxon' // the only import from luxon
[...]

		console.log(Duration.fromMillis(100000))
```

instead of above, it gives me a very similar picture:
![image](https://user-images.githubusercontent.com/182451/119350595-d1c36e80-bc9f-11eb-9fe6-be748c733fb9.png)

There is almost no difference.

If it's matter, I use webpack.

**Comment by JesusTheHun** [comment]
I was thinking about moving from `dayjs` which is foresaken to Luxon but I this tree shaking thing is a serious issue. Can maintainers speak to state whether or not tree shaking is possible and how to plan to improve it in the future ?
Are you working on this alone @icambron ?

**Comment by ralfstory** [comment]
> I was thinking about moving from `dayjs` which is foresaken to Luxon but I this tree shaking thing is a serious issue. Can maintainers speak to state whether or not tree shaking is possible and how they plan to improve it in the future ?

@JesusTheHun Can you explain what you mean by dayjs being forsaken?

**Comment by JesusTheHun** [comment]
> @JesusTheHun Can you explain what you mean by dayjs being forsaken?

Several core issues regarding Typescript support and ES6 are simply ignored for more than a year. PR for casual stuff are still merged but clearly the maintainer has other priorities - which is fine I mean it's FOSS but I'm looking for something robust for the years to come ; I would rather pay commercial support than having a lib already obsolete at the time of adoption.

**Comment by tamusjroyce** [comment]
In 2022, ESM imports were introduced. Would this allow tree shaking?

I find lots of things out there saying Luxon does not support tree shaking with webpack (and now rspack)

**Comment by Bessonov** [comment]
Looking at the [published code](https://www.npmjs.com/package/luxon?activeTab=code), I believe tree shaking should work. However, due to many code pieces being used by other code pieces and the extensive use of classes (both are not inherently bad!), the benefits of tree shaking are very limited:
```javascript
import { DateTime } from 'luxon'
console.log(DateTime.now())
```
![image](https://github.com/moment/luxon/assets/182451/52ebba9b-6d19-47ec-8135-80f3952de799)

**Comment by hatsantos** [comment]
Any progress on better tree shaking for the near future?

**Comment by onx2** [comment]
I would also like to see tree shaking in this package as it really seems promising and the API looks really nice. I can verify that as of the time of this post, tree shaking is effectively non-existent -- `import {DateTime} from "luxon"` leads to seemingly everything being added. Is it just the case that we must choose no tree shaking if we want a simpler API instead of something like date-fns?


<img width="862" alt="image" src="https://github.com/moment/luxon/assets/23039725/23e9342c-faff-4205-92cb-63cd80fff1b8">

**Comment by olejorgenb** [comment]
Yeah, from my recent attempt at understanding the state of tree shaking it appears to me that only module symbol level shaking work reliably? I found some tickets about class level shaking for web-pack, but not definitive documentation. Further more it seems that shaking does not even work when importing a full module "object"?

Which is IMO a shame, since this likely shape the whole javascript library ecosystem to use a certain design primarily because it's "shakeable" - instead of using a design that is optimal for the problem... It also means that javascript modules become very top-symbol heavy since you have to do the `import { ... } from '...'; or import singleSymbol from '...'` instead of `import moduleAsNamespace from '...'` to make shaking work.
